keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,399 @@
+- Immediate feedback from the K8s API if the user configures a Service with mixed protocol set but without this annotation
+Con:
+- Additional configuration task for the user
+- Must be executed for each and every Service definitions that are to define different protocols for the same LB
+
+#### Merging Services in CPI
+
+This one is not really a classic option control mechanism. The idea comes from the current practice implemented in MetalLB: https://metallb.universe.tf/usage/#ip-address-sharing
+The same works in GCE as a workaround.
+
+I.e. if a cloud provider wants to support this feature the CPI must have a logic to apply the Service definitions with a common key value (for example loadBalancerIP) on the same LoadBalancer instance. If the CPI does not implement this support it will work as it does currently.
+Pro:
+- the cloud provider can decide when to support this feature, and until that it works as currently for this kind of config
+- the restriction on mixed protocols can remain in the K8s API - i.e. there is no K8s API change, and the corresponding risks are mitigated
+Con:
+- the users must maintain two Service instances
+- Atomic update can be a problem - the key must be such a Service attribute that cannot be patched
+
+#### Proposed solution
+
+In the first release: 
+ - a feature flag shall control whether new loadbalancer Services with mixed protcol configuration can be created or not
+ - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
+ - the update of such Services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still validate the Service objects to check whether any of those has mixed protocol configuration and if the CPI cannot support such Service for any reasons the CPI shall indicate that clearly to the user, and the CPI should not create Cloud LB resources for the "accepted part" of the Service. The point is, that the system (Service - CPI implementation - cloud LB) must be in a consitent state in the end, and the user must have meaningful information about that state. For example, if the Service is not accepted by the CPI implementation this fact shall be available for the user, either as a new field in `Service.status.loadBalancer` or as a new Event (or both).
```

Comment:
Let's tighten this spec.

1) Once the apiserver allows mixed-protocol Service specs, each CPI has to decide if and how to support it.
2) If a CPI chooses to support mixed-protocol, everything is good
3) If a CPI can't support mixed-protocol, should they a) create the ones the support; b) fail the whole thing?
4) If a CPI can't support mixed-protocol, they must log that in new status fields (need a proposal for schema) and they should deliver an event

Reply:
Unfortunately, if a CPI does not support mixed-protocol LB and doesn't set status (e.g. has not been updated), there's not much we can do.

Reply:
Fixed. Tried to be more straight about the requirements. Also added the schema proposal.

Reply:
I'm exploring how much work is involved to have the CPI setup a validating webhook for Services Type=LB, so we can have provider-specific validation of port protocols. 

Hopefully it's available around the same time we lift this validation but don't think we need to block on it.